Add taxonomy block using SelectTree

nathanss opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This creates a "Hierarchical Taxonomy" block, which will be firstly used for WooCommerce Brands plugin, and also replace woocommerce/product-category-field block in the future.

Part of https://github.com/woocommerce/woocommerce/issues/39206 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install the "Custom Post Type UI" plugin by WebDevStudios
2. Set up some custom taxonomies.
    - Make sure to attach it to the Products post type
3. Add code similar to the following to a plugin/mu-plugin file, to add the properties to the REST API response and request:

```php
function YOUR_PREFIX_rest_api_prepare_custom1_to_product( $response, $post ) {
	$post_id = $post->get_id();

	if ( empty( $response->data[ 'custom1' ] ) ) {
		$terms = [];

		foreach ( wp_get_post_terms( $post_id, 'custom-taxonomy' ) as $term ) {
			$terms[] = [
				'id'   => $term->term_id,
				'name' => $term->name,
				'slug' => $term->slug,
			];
		}

		$response->data[ 'custom1' ] = $terms;
	}

	return $response;
}

add_filter( 'woocommerce_rest_prepare_product_object', 'YOUR_PREFIX_rest_api_prepare_custom1_to_product', 10, 2 );

function YOUR_PREFIX_rest_api_add_custom1_to_product( $product, $request, $creating = true ) {
	$product_id = $product->get_id();
	$params     = $request->get_params();
	$custom1s     = isset( $params['custom1'] ) ? $params['custom1'] : array();

	if ( ! empty( $custom1s ) ) {
		if ( $custom1s[0]['id'] ) {
			$custom1s = array_map(
				function ( $custom1 ) {
					return absint( $custom1['id'] );
				},
				$custom1s
			);
		} else {
			$custom1s = array_map( 'absint', $custom1s );
		}
		wp_set_object_terms( $product_id, $custom1s, 'custom-taxonomy' );
	}
}

add_filter( 'woocommerce_rest_insert_product_object', 'YOUR_PREFIX_rest_api_add_custom1_to_product', 10, 3 );
```

4. Add a block for the taxonomy property, using code similar to the following:
```php
$section_fields[] = [
	'woocommerce/taxonomy-field',
	[
		'slug'               => 'custom-taxonomy',
		'property' => 'custom1',
		'label'                                  => __( 'Custom Taxonomy', 'woocommerce' ),
		'createTitle'                        => __( 'Create new custom taxonomy', 'woocommerce' ),
	]
];
```

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

nathanss committed:
Create Hierarchical Taxonomy block

nathanss committed:
Add changelog

nathanss committed:
Update CSS and remove more references to 'brands'

mattsherman commented on packages/js/product-editor/src/blocks/hierarchical-taxonomy/edit.tsx:
```diff
@@ -0,0 +1,159 @@
+	const [ selectedEntries, setSelectedEntries ] = useEntityProp< Taxonomy[] >(
+		'postType',
+		'product',
+		taxonomyNamePlural
```
I think this is actually not necessarily the plural version of the taxonomy name, but rather the name of the property used in the REST API.

mattsherman commented on packages/js/product-editor/src/blocks/hierarchical-taxonomy/editor.scss:
```diff
@@ -0,0 +1,16 @@
+			margin-top: 36px;
+			display: flex;
+			flex-direction: row;
+			gap: 8px;
```
Can use `$gap-smaller`

nathanss committed:
Improve error message UX

mattsherman commented on packages/js/product-editor/src/blocks/hierarchical-taxonomy/editor.scss:
```diff
@@ -0,0 +1,16 @@
+		overflow: visible;
+
+		&__buttons {
+			margin-top: 36px;
```
Can use `$gap-larger`

mattsherman reviewed (changes requested):
Nice start on this @nathanss !

I need to take a closer look, but some initial thoughts...

* Some testing instructions that don't depend on the brands PR/plugin would be good
* Relatedly, it would be good to verify that this works with other custom taxonomies (I tried creating some with the "Custom Post Type UI" plugin, but I don't know why they don't seem to be appearing in the REST API responses (which I think might be the problem).
* It would be ideal if this block automatically handled both hierarchical and non-hierarchical taxonomies. I would think that should be possible without too much hassle.

nathanss committed:
Use css variables

nathanss committed:
Name block attribute

nathanss commented:
@mattsherman thx for the review!

> Some testing instructions that don't depend on the brands PR/plugin would be good

I can't think of a good way of doing this. I don't know how using `useEntityProp` would work in another context, like storybook, for example. One other way would be to migrate the category field. But ultimately I think the best way of testing this is with the brands plugin for now, since it is the first use case. Do you have any suggestions?

> Relatedly, it would be good to verify that this works with other custom taxonomies (I tried creating some with the "Custom Post Type UI" plugin, but I don't know why they don't seem to be appearing in the REST API responses (which I think might be the problem).

You're right, we need to have the property in the REST API. AFAIK we don't have any other property on the product to test this on for now.

> It would be ideal if this block automatically handled both hierarchical and non-hierarchical taxonomies. I would think that should be possible without too much hassle.

I will work on that.

nathanss committed:
Add support for non-hierarchical taxonomies

mattsherman commented:
> @mattsherman thx for the review!
> 
> > Some testing instructions that don't depend on the brands PR/plugin would be good
> 
> I can't think of a good way of doing this. I don't know how using `useEntityProp` would work in another context, like storybook, for example. One other way would be to migrate the category field. But ultimately I think the best way of testing this is with the brands plugin for now, since it is the first use case. Do you have any suggestions?

Ideally, I think testing one PR should not depend on changes in another PR. I have figured out how to test this without the Brands plugin (which is needed to test the non-hierarchical case too).

> > Relatedly, it would be good to verify that this works with other custom taxonomies (I tried creating some with the "Custom Post Type UI" plugin, but I don't know why they don't seem to be appearing in the REST API responses (which I think might be the problem).
> 
> You're right, we need to have the property in the REST API. AFAIK we don't have any other property on the product to test this on for now.

I figured out how to accomplish this.

> > It would be ideal if this block automatically handled both hierarchical and non-hierarchical taxonomies. I would think that should be possible without too much hassle.
> 
> I will work on that.

Thanks! 

Okay, here is an approach I have figured out to test this easily...

1. Install the "Custom Post Type UI" plugin by WebDevStudios (it's free; you really don't need it though... you could just register a taxonomy via code... I just used this to save a little bit of time/typing/learning when trying to get things working)
2. Set up some custom taxonomies.
    - "Taxonomy Slug" is our "taxonomyName"
    - Make sure to attach it to the Products post type
3. Add code similar to the following to a plugin/mu-plugin file, to add the properties to the REST API response:

```php
function YOUR_PREFIX_rest_api_add_custom1_to_product( $response, $post ) {
	$post_id = $post->get_id();

	if ( empty( $response->data[ 'custom1' ] ) ) {
		$terms = [];

		foreach ( wp_get_post_terms( $post_id, 'custom-taxonomy' ) as $term ) {
			$terms[] = [
				'id'   => $term->term_id,
				'name' => $term->name,
				'slug' => $term->slug,
			];
		}

		$response->data[ 'custom1' ] = $terms;
	}

	return $response;
}

add_filter( 'woocommerce_rest_prepare_product_object', 'YOUR_PREFIX_rest_api_add_custom1_to_product', 10, 2 );
```

4. Add a block for the taxonomy property, using code similar to the following (using pre-API technique just because we don't have template registration in quite yet):

```php
$section_fields[] = [
	'woocommerce/taxonomy-field',
	[
		'taxonomyName'               => 'custom-taxonomy',
		'taxonomyNameOnEntity' => 'custom1',
		'label'                                  => __( 'Custom Taxonomy', 'woocommerce' ),
		'createTitle'                        => __( 'Create new custom taxonomy', 'woocommerce' ),
	]
];
```

mattsherman commented on packages/js/product-editor/src/blocks/taxonomy/block.json:
```diff
@@ -0,0 +1,43 @@
+			"type": "string",
+			"__experimentalRole": "content"
+		},
+		"taxonomyNameOnEntity": {
```
How about `property`? That's what other generic fields like our checkbox and radio button blocks use.

mattsherman commented on packages/js/product-editor/src/blocks/taxonomy/block.json:
```diff
@@ -0,0 +1,43 @@
+			"type": "boolean",
+			"default": false,
+			"__experimentalRole": "content"
+		}
```
I don't think the consumer should set this. We can determine this by querying for details on the taxonomy.

mattsherman commented on packages/js/product-editor/src/blocks/taxonomy/block.json:
```diff
@@ -0,0 +1,43 @@
+	"keywords": [ "taxonomy"],
+	"textdomain": "default",
+	"attributes": {
+		"taxonomyName": {
```
This is the taxonomy slug. Maybe we should call it `taxonomySlug`.

mattsherman reviewed (changes requested):
Excellent iteration on this so far! Overall, things appear to be working pretty well with the latest changes. I will do a more rigorous final test and walk through the code once the remaining changes are made.

nathanss committed:
Rename attributes

nathanss committed:
Fetch "hierarchical" from taxonomy data

nathanss commented:
@mattsherman Awesome review, thanks a lot.

I updated the code with the suggested changes. I also updated the PR title and description, since it evolved a lot since the beginning. Can you check again?

nathanss committed:
Update changelog

nathanss committed:
Remove 'hierarchical' attribute

mattsherman commented on packages/js/product-editor/src/blocks/taxonomy/block.json:
```diff
@@ -0,0 +1,38 @@
+	"name": "woocommerce/taxonomy-field",
+	"title": "Taxonomy",
+	"category": "widgets",
+	"description": "A block that displays a taxonomy field, allowing to search, select, and create new entries",
```
Suggestion...

```
"A block that displays a taxonomy field, allowing searching, selection, and creation of new items"
```

mattsherman commented on packages/js/product-editor/src/blocks/taxonomy/create-taxonomy-modal.tsx:
```diff
@@ -0,0 +1,190 @@
+					/>
+				</BaseControl>
+				{ hierarchical && (
+					<SelectTree
```
This select tree does not show parents nested, as the categories modal does. It shows them in a flat list, which is confusing....

<img width="680" alt="Screenshot 2023-08-30 at 19 53 31" src="https://github.com/woocommerce/woocommerce/assets/2098816/ed666287-7f69-48ba-9011-7814f5191870">

<img width="680" alt="Screenshot 2023-08-30 at 19 54 20" src="https://github.com/woocommerce/woocommerce/assets/2098816/59d6dd76-8f56-4348-b664-245156e6d434">

mattsherman commented on packages/js/product-editor/src/blocks/taxonomy/edit.tsx:
```diff
@@ -0,0 +1,185 @@
+	const hierarchical = useSelect( ( select ) => {
+		const taxonomyDetails = select( 'core' ).getTaxonomy(
+			attributes.slug
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Can we avoid this?

mattsherman commented:
Note: forgot that we have to handle setting the taxonomy in the REST request as well... updated the testing instructions to account for that.

mattsherman reviewed (changes requested):
Testing seems to work pretty well with the `brands` taxonomy, as well as two custom taxonomies I created, one hierarchical, one flat (after I added the handling to save the taxonomy in the REST handler*).

The main issue I ran into during testing was that the parent dropdown in the modal does not show parents hierarchically.

\* I suppose that this generic taxonomy block only works out of the box if the custom taxonomy developer has implemented support for the REST API as is being done for brands and categories (same shape of data for the property). We'll have to document this. I think it is a reasonable limitation, and maybe one we can extend over time if there are other patterns.

nathanss committed:
Update description

nathanss committed:
Fix issue where hierarchy was not displayed in create modal

nathanss committed:
Refactor taxonomy details select

nathanss committed:
Add README to taxonomy block

mattsherman reviewed (approved):
Nice work on this, @nathanss -- latest changes look good! :shipit:
